Sources/Purchasing/StoreKit2/Win-Back Offers/WinBackOfferEligibilityCalculator.swift

```diff
@@ -0,0 +1,135 @@
+extension WinBackOfferEligibilityCalculator: WinBackOfferEligibilityCalculatorType {
+
+    func eligibleWinBackOffers(forProduct product: StoreProduct) async throws -> [WinBackOffer] {
+        return try await self.calculateEligibleWinBackOffers(forProduct: product)
+    }
+
+}
+
+// MARK: - Implementation
+@available(iOS 18.0, macOS 15.0, tvOS 18.0, watchOS 11.0, visionOS 2.0, *)
+extension WinBackOfferEligibilityCalculator {
+
+    @available(iOS 18.0, macOS 15.0, tvOS 18.0, watchOS 11.0, visionOS 2.0, *)
+    private func calculateEligibleWinBackOffers(
+        forProduct product: StoreProduct
+    ) async throws -> [WinBackOffer] {
+
+        #if compiler(>=6.0)
+        guard self.systemInfo.storeKitVersion.isStoreKit2EnabledAndAvailable else {
+            throw ErrorUtils.featureNotSupportedWithStoreKit1Error()
+        }
+
+        let eligibleWinBackOfferIDs: [String] = await self.calculateEligibleWinBackOfferIDs(forProduct: product)
+        guard !eligibleWinBackOfferIDs.isEmpty else { return [] }
+
+        guard let allWinBackOffersForThisProduct: [
+            Product.SubscriptionOffer
+        ] = product.sk2Product?.subscription?.winBackOffers else {
+            // StoreKit.Product.SubscriptionInfo is nil if the product is not a subscription
+            return []
+        }
+
+        let eligibleWinBackOffers = allWinBackOffersForThisProduct
+            // 1. Filter out the offers that the subscriber is not eligible for
+            .filter({
+                Set(eligibleWinBackOfferIDs).contains($0.id)
+            })
+            // 2. Convert the eligible offers to StoreProductDiscounts for us to use
+            .compactMap({
+                StoreProductDiscount(sk2Discount: $0, currencyCode: product.currencyCode)
+            })
+            // 3. Convert the StoreProductDiscounts to WinBackOffer objects
+            .map({
+                WinBackOffer(discount: $0)
+            })
+
+        return eligibleWinBackOffers
+        #else
+        return []
+        #endif
+    }
+
+    @available(iOS 18.0, macOS 15.0, tvOS 18.0, watchOS 11.0, visionOS 2.0, *)
+    private func calculateEligibleWinBackOfferIDs(forProduct product: StoreProduct) async -> [String] {
+        #if compiler(>=6.0)
+        guard let statuses = try? await product.sk2Product?.subscription?.status, !statuses.isEmpty else {
+            // If StoreKit.Product.subscription is nil, then the product isn't a subscription
+            // If statuses is empty, the subscriber was never subscribed to a product in the subscription group.
+            return []
+        }
+
+        let purchasedSubscriptionStatuses = statuses.filter({
+            switch $0.transaction {
+            case .unverified:
+                return false
+            case .verified(let transaction):
+                // Intentionally exclude transactions acquired through family sharing
+                return transaction.ownershipType == .purchased
```

Comment:
Technically, you can only be subscribed to a product once, so there must be at most 1 renewalInfo instance that is NOT a family shared one. So maybe we could simplify the logic of the rest of the function by just finding the first status transaction with `transaction.ownershipType == .purchased`, like Apple does in their sample code:

```
let purchasedStatus = statuses.first {
    $0.transaction.unsafePayloadValue.ownershipType == .purchased
}
```

https://developer.apple.com/videos/play/wwdc2024/10110/

Reply:
Yeah, this is where I thought I might be able to do a bit better than Apple's example in the event where there were multiple renewalInfos.

Is there any chance that there would be more than one renewalInfo for a single product ID if a user subscribes, churns, and then re-subscribed? I'm not seeing many details about this in the docs.

I've implemented this change [here](https://github.com/RevenueCat/purchases-ios/pull/4435) and will merge it in to this PR if we think we're safe in the case I mentioned above. 👍

Reply:
The documentation does not explicitly state that, but it says:

> The array can have more than one subscription status if your subscription supports Family Sharing.

Which I'd argue highly suggest the array can normally only have one status (the purchased one) and potentially more only if it supports family sharing.

Given that, plus Apple's sample code makes me very confident there will always be just one status with `ownershipType == .purchased`

https://developer.apple.com/documentation/storekit/product/subscriptioninfo/3822296-status